rfcs/0003-mutations.md

```diff
@@ -0,0 +1,317 @@
+
+## Urgency
+
+This is urgent from a developer experience point of view. With this addition, it eliminates the need for protocol developers to manually wrap GraphQL query interfaces alongside developer-friendly write functions. Additionally, mutations provide a solution for optimistic UI updates, which is something dApp developers have been seeking for a long time (see [here](https://github.com/aragon/nest/issues/21)). Lastly with the whole protocol now defined in GraphQL, existing application layer code generators can now be used to hasten dApp development ([some examples](https://dev.to/graphqleditor/top-3-graphql-code-generators-1gnj)).
+
+## Terminology
+
+* _Mutations_: Collection of mutations.  
+* _Mutation_: A GraphQL mutation.  
+* _Mutations Schema_: A GraphQL schema that defines a `type Mutation` that contains all mutations. Additionally, this schema can define other types to be used by the mutations, such as `input` and `interface` types.  
+* _Mutations Manifest_: A YAML manifest file that is used to add mutations to an existing subgraph manifest.  
+* _Mutation Resolvers_: Code module that contains all resolvers.  
+* _Resolver_: Function that is used to execute a mutation.  
+* _Mutation State_: The state of a mutation being executed. It's passed to the resolver through the mutation context.  
+* _Mutation Context_: A context object that's created for every mutation that's executed. It's passed as an argument to the resolver.  
+* _Config_: Collection of config properties required by the mutation resolvers.  
+* _Config Property_: A single property within the config (ex: ipfs, ethereum, etc).  
+* _Config Generator_: A function that takes a config value, and returns a config property. For example, "localhost:5001" as a config value gets turned into a new IPFS client by the config generator.
+* _Config Value_: An initialization value that's passed into the config generator. This config value is provided by the dApp developer.
+* _Optimistic Response_: A response given to the dApp that predicts what the outcome of the mutation's execution will be. If it is incorrect, it will be overwritten with the actual result.  
+
+## Detailed Design
+
+The sections below illustrate how a developer would add mutations to an existing subgraph, and then add those mutations to a dApp.
+
+### Mutations Manifest
+
+The subgraph manifest (`subgraph.yaml`) now has an extra property named `mutations` which is the mutations manifest.
+
+`subgraph.yaml`
+```yaml
+specVersion: ...
+...
+mutations:
+  repository: https://npmjs.com/package/...
+  schema:
+    file: ./mutations/schema.graphql
+  resolvers:
+    apiVersion: 0.0.1
+    kind: javascript/es5
+    file: ./mutations/index.js
+dataSources: ...
+...
+```
+
+Alternatively, the mutation manifest can be external like so:  
+`subgraph.yaml`
+```yaml
+specVersion: ...
```

Comment:
It may be worth noting that the `specVersion` here must match the `specVersion` of the main subgraph manifest. That way we tie the two together, which will make our live easier whenever we need to bump this version.

Reply:
Our validation in graph-cli should catch mismatches.

Reply:
I was thinking we would just omit the specVersion within the mutations.yaml (if it's separate), as it's meaningless without the subgraph.yaml. If specVersion is only within subgraph.yaml, a mismatch would never occur.

Reply:
That's true, although if multiple subgraphs use the same mutations manifest, and one of them uses a different specVersion, we'd want to know if there is a mismatch. I think it's worth keeping.

Reply:
Okay sounds like a plan :+1: